### PR TITLE
Minor cleanups in dl4j-scaleout

### PR DESCRIPTION
**Pull request by huitseeker**
DO NOT MERGE (this touches a Spark section that is brittle atm, this branch, while relatively benign on its own, would only confuse things)

### CODE REVIEW

**Comment by huitseeker** [comment]
c8bc80a29a7a785c17462423fcea4e0e409d27b0 passes tests on my side (FINALLY!), removing WIP

**Comment by huitseeker** [comment]
Jenkins test this please

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Created by huitseeker on 2/15/17.
+ */
+public class SparkListenable implements Serializable {
+    protected static final Logger log = LoggerFactory.getLogger(SparkComputationGraph.class);
+    protected TrainingMaster trainingMaster;
```
Comment (agibsonccc) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java]: Lombok @Slf4j instead?

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @param statsStorage Stats storage router to place the results into
+     * @param listeners    Listeners to set
+     */
+    public void setListeners(StatsStorageRouter statsStorage, Collection<? extends IterationListener> listeners) {
+        //Check if we have any RoutingIterationListener instances that need a StatsStorage implementation...
+        StatsStorageRouterProvider routerProvider = null;
+        if(listeners != null ){
+            for(IterationListener l : listeners){
+                if(l instanceof RoutingIterationListener){
+                    RoutingIterationListener rl = (RoutingIterationListener)l;
```
Comment (agibsonccc) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java]: spacing on braces?

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/BaseDataSetIterator.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException();
+    }
+
+    private void preloadDataSet(){
+        preloadedDataSet = load(iter.next());
+        totalOutcomes = preloadedDataSet.getLabels().size(1);
+        inputColumns = preloadedDataSet.getFeatureMatrix().size(1);
+        batch = preloadedDataSet.numExamples();
```
Comment (agibsonccc) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/BaseDataSetIterator.java]: Need to be careful here. We can't always assume classification problems. Maybe use -1 (last dimension) instead?

**Comment by saudet** [comment]
Start Jenkins

**Comment by huitseeker** [comment]
Start Jenkins

**Comment by agibsonccc** [comment]
Could we get this merged within the next few days? We'd like to push for a release this week.

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/** Created by huitseeker on 2/15/17. */
+@Slf4j
+public class SparkListenable implements Serializable {
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java]: This shouldn't implement serializable. TrainingMasters aren't serializable (nor should they be), and ~~~instances of these (SparkDl4jMultiLayer and SparkComputationGraph) aren't either.~~~

Edit: turns out SparkDl4jMultiLayer and SparkComputationGraph *are* serializable. But IMO they shouldn't be (again, TrainingMaster instances aren't, and we don't want to use Java serialization for the MultiLayerNetwork and ComputationGraph fields anyway). Implementing serializable suggests to users that java serialization is valid for persisting these objects (it's not).

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  public void setListeners(
+      StatsStorageRouter statsStorage, Collection<? extends IterationListener> listeners) {
+    //Check if we have any RoutingIterationListener instances that need a StatsStorage implementation...
+    StatsStorageRouterProvider routerProvider = null;
+    if (listeners != null) {
+      for (IterationListener l : listeners) {
+        if (l instanceof RoutingIterationListener) {
+          RoutingIterationListener rl = (RoutingIterationListener) l;
+          if (rl.getStorageRouter() == null) {
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java]: Condition should be ```statsStorage == null```; whereas ```rl.getStorageRouter()``` being null is totally fine.
(And yes, I realize that's just a refactoring of my code :))

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            log.warn(
+                "RoutingIterationListener provided without providing any StatsStorage instance. Iterator may not function without one. Listener: {}",
+                l);
+          } else if (!(rl.getStorageRouter() instanceof Serializable)) {
+            //Spark would throw a (probably cryptic) serialization exception later anyway...
+            throw new IllegalStateException(
+                "RoutingIterationListener provided with non-serializable storage router");
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/SparkListenable.java]: Might as well log the RoutingIterationListener and StatsStorageRouter classes, if you make the previous/above changes to aid debugging?

**Comment by huitseeker** [comment]
Done with PR comments, but please have a look at https://github.com/deeplearning4j/deeplearning4j/issues/2909 before merging.

**Pull request merged**